src/lib/router.ts

```diff
@@ -100,31 +111,58 @@ export function updateProjectRoute(
   }
 }
 
-export const push = (newRoute: Route): void => {
+const historyExecutor = mutexExecutor.create();
+
+async function setHistory(history: Route[]): Promise<void> {
+  showLoadingBar.set(true);
+  const newRoute = history.slice(-1)[0];
+
+  // TODO: is there a way to somehow abort the previous fetches if we
+  // receive new navigation requests?
+  const loadedRoute = await historyExecutor.run(async () => {
+    if (newRoute.resource === "seeds") {
+      return await loadSeedRoute(newRoute.params);
+    } else if (newRoute.resource === "home") {
+      return await loadHomeRoute();
+    } else {
+      return newRoute;
+    }
```

Comment:
This code is gone grow with every route you add loading for. I would suggest you split responsibilities between concrete and abstract loading of routes. This means you only call a function here 
```typescript
const loadedRoute = await historyExecutor.run(() => loadRoute(newRoute))
```

And then implement the function in `router/defnition.ts`.
```typescript
export async function loadRoute(route: Route): Promise<LoadedRoute> {
  if (route.resource === "seeds") {
    return await loadSeedRoute(route.params);
  } else if (route.resource === "home") {
    return await loadHomeRoute();
  } else {
    return route;
  }
}
```

One goal of the improved separation of responsibilities is that this module should not depend on the shape of `Route` and `LoadedRoute`. If you change either of the two you should not need to touch this module. This means as a follow-up `pathToRoute` and other functions should be extracted.

Reply:
138cc472f51ecffe94cc82057cc2f27338af68f8

Reply:
> Not sure I follow entirely. So you mean the router.ts module should only have stuff that is concerned with the browser history handling? And we move everything else out into separate modules?

I wouldn’t say that `router.ts` should be limited to browser history handling. But I am saying that code that depends on the specific routes should live somewhere else. So at some point `routeToPath` and `pathToRoute` should move somewhere else, too.

The idea behind this is that the routing code should be generic with respect to the concrete shape of the routes. This makes it easier to understand the routing code. And, as I said above, changes to the routes like adding a route or changing parameters would not lead to changes to the routing system.